Improved OperationSelector performance

Donnerbart opened the pull request:
Improved `OperationSelector` performance by creating an `OperationSelectorBuilder` which constructs an  `OperationSelector` instance with fast data structures.

To prevent contention on the `Random` instance a new instance is created for each `Worker` in all tests which use `OperationSelector`.

Should fix #485.

jerrinot commented on stabilizer/src/main/java/com/hazelcast/stabilizer/worker/selector/OperationSelectorBuilder.java:
```diff
@@ -0,0 +1,97 @@
+     *
+     * @return instance of OperationSelector
+     */
+    public OperationSelector<T> getOperationSelector() {
```
what about something like build() ? 
I believe that's a common name for a builder pattern.

Donnerbart commented:
Changed probability values for `OperationSelector` from 0.0 to 1.0 to percentage (0 to 100).

I think this makes it easier for us to calculate probabilities and to refactor old tests, which use percentage more often than a probability of max 1.0.

jerrinot commented:
:+1:

Donnerbart commented:
Missed a line in `OperationSelector.addOperationRemainingProbability()` to change from subtracting to adding the total probability. Added a test case for this method.

pveentjer commented on stabilizer/src/main/java/com/hazelcast/stabilizer/worker/selector/OperationSelector.java:
```diff
@@ -0,0 +1,52 @@
+     */
+    public T select() {
+        int chance = random.nextInt(100);
+        for (Map.Entry<T, Integer> entry : operations.entrySet()) {
```
You can come up with a better structure to find a random operation. Now you need to iterate over 50 items on average to find a good one; this in my book is very bad practice.

Tip: use an array of T with size 100 and the generated random is the index in this array. Now you don't need to loop over anything.

pveentjer replied:
So for discussion sake, lets assume we have 10 instead of 100 chances, and you will have 80% read and 20% put, then the array would look like:
[put, put, read, read, read, read, read, read, read, read]

If you want you can randomize it, but the random index will already do that for you.

Donnerbart replied:
That's a good catch, missed that when switching to percentage (which I still think is the better API). Since native array of <Generic> won't work I'll find another solution, e.g. with `Object[]`.

pveentjer commented on stabilizer/src/main/java/com/hazelcast/stabilizer/worker/selector/OperationSelectorBuilder.java:
```diff
@@ -0,0 +1,86 @@
+ *
+ * @param <T> enum of operations
+ */
+public class OperationSelectorBuilder<T extends Enum<T>> {
```
More documentation is needed. OperationSelectorBuilders are not threadsafe and you also need to explain that each worker-thread should get its own instance since the OperationSelector itself isn't threadsafe.

Donnerbart replied:
Why isn't the `OperationSelector` thread-safe? It has just immutable fields and should be safely constructed. You may want to create multiple instances to reduce contention on `Random`, but it shouldn't be necessary.

pveentjer replied:
Random is 'not' threadsafe.

pveentjer replied:
Apart from that; every worker will get its own OperationSelector so therefore it doesn't need to be threadsafe. That is what you need to put in the documentation so that people know how to use the API correctly

Donnerbart replied:
I don't question that we should create an instance per worker and I will update the doc to recommend this. Nevertheless I really have in mind that `Random` is be thread-safe. I looked it up and it is in the docs since Java 7: http://docs.oracle.com/javase/7/docs/api/java/util/Random.html

So yes, maybe it will fail for all Hazelcast related projects on Java 6, so we should not relay on it.

pveentjer commented on stabilizer/src/main/java/com/hazelcast/stabilizer/worker/selector/OperationSelector.java:
```diff
@@ -0,0 +1,52 @@
+     * Select an operation according to configured probabilities.
+     * <p/>
+     * Selects a random operation, based on configured probabilities. If probabilities are smaller than 100 no operation may be
+     * selected. If a default operation is configured it is returned in that case. Otherwise <tt>null</tt> is returned.
```
How should a user deal with null? I don't see a good usecase for this. I think an exception is better fitting.

Donnerbart replied:
It's just the old behavior. I would change the `OperationSelectorBuilder` to throw an exception (instead of logging a warning), if the user missed to pass enough operations (fail fast). I think that's better than doing this during `OperationSelector.select()`.

Donnerbart commented:
All review issues should be fixed.

pveentjer commented on docs/TESTS.md:
```diff
@@ -59,8 +62,7 @@ public class ExampleTest {
 
-        selector.addOperation(Operation.PUT, putProbability)
-                .addOperation(Operation.GET, getProbability);
+        operationSelectorBuilder.addOperation(Operation.PUT, putPercentage).addOperation(Operation.GET, getPercentage);
```
In the future (not now) we should be able to let the test framework take care of this. It can read the properties and build up the selector builder.

pveentjer commented on stabilizer/src/main/java/com/hazelcast/stabilizer/worker/selector/OperationSelector.java:
```diff
@@ -0,0 +1,35 @@
+    private final Object[] operations;
+
+    OperationSelector(Object[] operations) {
+        this.random = new Random();
```
Personal flavour; do whatever you like. If an object can be initialized in the class constructor, I do that. It keeps the normal constructor cleaner and shorter. Especially when you have some initialization logic in the normal constructor, this can be a good way to get rid of things that don't to be there.

pveentjer commented on stabilizer/src/main/java/com/hazelcast/stabilizer/worker/tasks/AbstractWorkerTask.java:
```diff
@@ -37,7 +38,7 @@
 
     public AbstractWorkerTask() {
-        this.selector = createOperationSelector();
+        this.selector = createOperationSelectorBuilder().build();
```
I don't think this belongs in the abstract worker. I want to prevent adding too much framework and certainly want to prevent forcing a user to implement methods if he doesn't care about them.

Afaik the OperationSelector can also be initialized when the worker is created in the concrete subclass

```
class FooWorker extends AbstractWorker{
     OperationSelector<FoOOPeration> selector = selectorBuilder.newInstance();
     ....
}
```

Donnerbart replied:
I thought this whole approach should be the lazy bum way, so the `AbstractWorkerTask` should do all the things which are needed for a standard test. This will use an `OperationSelector`, so you should guide the user to provide a builder for it. The `doRun()` receives the selected operation, so the `select()` is also done by the `AbstractWorkerTask`, and I think that is very convenient if you want to write a standard test. If you want to do something funky there is still `@Run` where you can setup everything manually.

pveentjer replied:
Ok. Makes sense.

Would it not make sense to pass in the selector builder into the constructor of the worker? This way you don't need to override a method.

Donnerbart replied:
I can give that a try. That would solve the problem that we create several builders at the moment, which is technically not necessary.

Donnerbart commented:
PR updated with latest review comments.

Donnerbart commented:
Rebased with master and adapted all new tests.

Donnerbart commented:
*ping*

Donnerbart commented:
Danny commented that using percentages for probabilities is a very uncommon thing. So should I revert that part of the PR? Should we nevertheless keep the internal array and the max. granularity of 100? It's such a nice and fast internal implementation. WDYT?

Donnerbart commented:
Changed the probability back to a double between 0.0 and 1.0, but kept the array for the `OperationSelector`. I think this is the best compromise between an intuitive API and a good underlying structure for performance. The maximum precision is 3 (so 0.001 as minimal value). I also added some test for missing edge cases and parameters.

![My work here is done]
(http://cdn.meme.am/instances/500x/53368971.jpg)
